File: document/Algorithm/sort/python/selection_sort.py
"""
Selection Sort
- Stable: No (unless modified)
- In-place: Yes
- Time: O(n^2)
- Space: O(1)
"""

def selection_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

if __name__ == "__main__":
    data = [64, 25, 12, 22, 11]
    print("before:", data)
    print("after :", selection_sort(data))

File: document/Algorithm/sort/java/SelectionSort.java
/*
Selection Sort
- Stable: No
- In-place: Yes
- Time: O(n^2)
- Space: O(1)
*/
public class SelectionSort {
    static void printArr(int[] a) {
        for (int i = 0; i < a.length; i++) {
            if (i > 0) System.out.print(" ");
            System.out.print(a[i]);
        }
        System.out.println();
    }
    public static void sort(int[] a) {
        int n = a.length;
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) if (a[j] < a[minIdx]) minIdx = j;
            if (minIdx != i) {
                int t = a[i]; a[i] = a[minIdx]; a[minIdx] = t;
            }
        }
    }
    public static void main(String[] args) {
        int[] a = {64,25,12,22,11};
        sort(a);
        printArr(a);
    }
}
